packages/ui/react-ui-table/src/model/table-sorting.ts

```diff
@@ -0,0 +1,124 @@
+//
+// Copyright 2025 DXOS.org
+//
+
+import { computed, type Signal, signal, type ReadonlySignal } from '@preact/signals-core';
+import orderBy from 'lodash.orderby';
+
+import { getValue, FormatEnum, TypeEnum } from '@dxos/echo-schema';
+import { formatForDisplay } from '@dxos/react-ui-form';
+import { type PropertyType, type FieldType, type ViewProjection } from '@dxos/schema';
+
+import { type BaseTableRow } from './table-model';
+import { type TableType } from '../types';
+
+export type SortDirection = 'asc' | 'desc';
+export type SortConfig = { fieldId: string; direction: SortDirection };
+
+/**
+ * Manages table sorting functionality through display and data index mapping.
+ *
+ * When a table is sorted, rows visually reorder but we need to maintain references
+ * to the original data positions. This is handled through two types of indices:
+ *
+ * - Display indices: The visual positions after sorting
+ * - Data indices: The original positions in the source data
+ *
+ * The _displayToDataIndex map maintains these relationships, enabling lookups
+ * between visual and source positions.
+ *
+ * Example mapping between display and data indices:
+ *
+ * Original Data:        Sorted Display:       Mapping: (Display → Data)
+ * ┌──────────────┐      ┌──────────────┐      ┌────────┐
+ * │ 0: "Banana"  │  =>  │ 0: "Apple"   │  =>  │ 0 => 1 │
+ * │ 1: "Apple"   │  =>  │ 1: "Banana"  │  =>  │ 1 => 0 │
+ * │ 2: "Cherry"  │  =>  │ 2: "Cherry"  │  =>  │ 2 => 2 │
+ * └──────────────┘      └──────────────┘      └────────┘
+ */
+export class TableSorting<T extends BaseTableRow> {
+  private readonly _displayToDataIndex = new Map<number, number>();
+  private readonly _sorting = signal<SortConfig | undefined>(undefined);
+  private readonly _rows: Signal<T[]>;
+  public readonly sortedRows: ReadonlySignal<T[]>;
+
+  constructor(
+    rows: Signal<T[]>,
+    private readonly _table: TableType,
+    private readonly _projection: ViewProjection,
+  ) {
+    this._rows = rows;
+    this.sortedRows = this.initialiseSortedRows();
+  }
+
+  private initialiseSortedRows(): ReadonlySignal<T[]> {
```

Comment:
nit: I tend to prefer the public api coming at the top of the class after the constructor